src/lucky/param_serializable.cr

```diff
@@ -0,0 +1,177 @@
+          {% end %}
+
+          if %val.nil?{% if metadata[:nilable] %} || %val == ""{% end %}
+            default_or_nil = {{ metadata[:type_declaration].value ? metadata[:type_declaration].value : nil }}
+            {% if metadata[:nilable] %}
+              if %val.nil? && default_or_nil.nil?
+                @{{ metadata[:type_declaration].var.id }} = nil
+              else
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: default_or_nil, param_key: %param_key_value)
+              end
+            {% else %}
+              if default_or_nil.nil?
+                raise Lucky::MissingParamError.new <<-ERROR
+                {{ @type }} is missing value for required param "{{ metadata[:type_declaration] }}"
+                ERROR
+              else
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: default_or_nil, param_key: %param_key_value)
+              end
+            {% end %}
+          else
+            {% if metadata[:is_serializable] %}
+            @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: {{ metadata[:core_type] }}.from_params(%val.as(Lucky::Params)), param_key: %param_key_value)
+            {% else %}
+              # NOTE: these come from Avram directly
+              %result = {{ metadata[:core_type] }}::Lucky.parse(%val)
+
+              if %result.is_a? Avram::Type::SuccessfulCast
+                @{{ metadata[:type_declaration].var.id }} = Lucky::PermittedParam({{ metadata[:type_declaration].type }}).new(name: {{ metadata[:type_declaration].var.stringify }}, value: %result.value.as({{ metadata[:type] }}), param_key: %param_key_value)
+              else
+                raise Lucky::InvalidParamError.new(
+                  param_name: "{{ metadata[:type_declaration].var.id }}",
+                  param_value: %val.to_s,
+                  param_type: "{{ metadata[:type] }}"
+                )
+              end
+            {% end %}
+          end
+        {% end %}
+      {% end %}
+    end
+
+    macro param(type_declaration, param_key = nil)
+      {% unless type_declaration.is_a?(TypeDeclaration) %}
+        {% raise "'param' expects a type declaration like 'name : String', instead got: '#{type_declaration}'" %}
+      {% end %}
+
+      {% is_nilable_type = type_declaration.type.resolve.nilable? %}
+      {% type = is_nilable_type ? type_declaration.type.types.reject(&.==(Nil)).first.resolve : type_declaration.type.resolve %}
+      {% is_array = type.name.starts_with?("Array") %}
+      # TODO: This probably breaks if the type is Array(String?)
+      {% core_type = is_array ? type.type_vars.first : type %}
```

Comment:
If the type is nilable (`String?`), then `type` is `String`, but if it's an `Array(String?)`, then `core_type` ends up being `String?` and not `String`.... I think that will cause some issues 🤔